Upgrade bleach to 2.1.4

ahmedaljazzar opened the pull request:
This PR allows the platform to use the version `2.1.4` of `bleach` library. This is needed as we're at OpenCraft using it in building the new [HTML XBlock](https://github.com/open-craft/xblock-html).
The change requires another upgrade for the same library which took place in [edx-ora2](https://github.com/edx/edx-ora2/pull/1129) and [RecommenderXBlock](https://github.com/edx/RecommenderXBlock/pull/5). Once these prs got approved, I can update the repo versions here instead of the ones I added for test purposes.

#### Testing instructions
1. Test ORA
    - [ ] Go to Studio and create a new ORA unit.
    - [ ] Add script tag by inspecting the textarea tinymce is using since ORA doesn't support html editing.
    - [ ] Make sure the saved text is bleached after saving the changes and is not reflected on LMS.
1. Test CAPA
    - [ ] Go to Studio and create a new CAPA unit.
    - [ ] Edit the text area directly and add your script or evil tag.
    - [ ] The module should not accept that change and shows the same behavior in master
1. Test Recommender XBlock
    - [ ] Test Recommender XBlock.

#### Dependencies
- [ ] [edx-ora2#1129](https://github.com/edx/edx-ora2/pull/1129)
- [ ] [RecommenderXBlock#5](https://github.com/edx/RecommenderXBlock/pull/5)

#### Notes
This PR has been created as a replacement of #18963

#### People involved
- @bradenmacdonald 
- @mduboseedx

edx-webhook commented:
Thanks for the pull request, @ahmedaljazzar! I've created [OSPR-2651](https://openedx.atlassian.net/browse/OSPR-2651) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

mduboseedx commented:
jenkins ok to test

mduboseedx commented:
@ahmedaljazzar Could you take a look at the failed tests?

ahmedaljazzar commented:
@mduboseedx Sure, all of them are related to Matlab so I'll go forward with check that functionality now.

bradenmacdonald commented:
@ahmedaljazzar Looks like there is still some issue with the recommender xblock integration tests

ahmedaljazzar commented:
@mduboseedx The tests are passing now and it's ready for another pass.

mduboseedx commented:
@jmbowman Do you think you could take a look at this? It is upgrading the bleach library

jmbowman commented on common/lib/capa/capa/util.py:
```diff
@@ -173,8 +173,8 @@ def sanitize_html(html_code):
     attributes = bleach.ALLOWED_ATTRIBUTES.copy()
     # Yuck! but bleach does not offer the option of passing in allowed_protocols,
```
As of bleach 1.5, this is no longer true: https://bleach.readthedocs.io/en/latest/clean.html#allowed-protocols-protocols .  Could you switch the code to pass the new `protocols` argument instead of mutating the default list?

jmbowman commented on requirements/edx/base.in:
```diff
@@ -93,7 +93,7 @@ futures ; python_version == "2.7"   # via django-pipeline, python-swift-client,
 gunicorn==17.5
 help-tokens
-html5lib==0.999                     # HTML parser, used for capa problems
+html5lib==1.0.1                     # HTML parser, used for capa problems
```
Since this is the current release, could you just remove the version constraint?  That would let us automatically notice and test new versions via `make upgrade`, while still keeping the version pinned in `base.txt`, etc.

jmbowman commented on requirements/edx/github.in:
```diff
@@ -101,4 +101,5 @@
 
 -e git+https://github.com/mitodl/edx-sga.git@3828ba9e413080a81b907a3381e5ffa05e063f81#egg=edx-sga==0.8.3
 -e git+https://github.com/open-craft/xblock-poll@add89e14558c30f3c8dc7431e5cd6536fff6d941#egg=xblock-poll==1.5.1
+-e git+https://github.com/open-craft/xblock-html.git@jazzar/initial-changes#egg=xblock-html==1.0.0
```
Was this deliberately added?  I don't see it reflected in the `*.txt` files, and I thought the main point of the PR was to upgrade bleach so this could be added later.

bradenmacdonald replied:
Good catch, `xblock-html` should not be included for now.

ahmedaljazzar replied:
Good catch! Committed by mistake when I was testing the upgraded bleacher with external XBlocks.

ahmedaljazzar commented:
@jmbowman Thanks for the review. I addressed all the comments above. This is ready for another pass.

jmbowman commented on requirements/edx/base.in:
```diff
@@ -93,7 +93,6 @@ futures ; python_version == "2.7"   # via django-pipeline, python-swift-client,
 glob2==0.3                          # Enhanced glob module, used in openedx.core.lib.rooted_paths
 gunicorn==17.5
 help-tokens
-html5lib==0.999                     # HTML parser, used for capa problems
```
I think we want to keep a line for html5lib, because it's a direct dependency of the capa code in this repository.  I was just suggesting that it shouldn't be pinned to an exact version here.  As we gradually get individual dependencies up to date, this file will become mostly a list of package names with no version constraints.

ahmedaljazzar replied:
Oh I see! Sorry about that.

ahmedaljazzar replied:
This is now fixed :D

ahmedaljazzar commented:
Awesome! Thanks @jmbowman

bradenmacdonald commented:
@jmbowman Would you be able to look at the PRs on those other repos which are ready for you review, so we can get this merged, or do those fall to someone else?

jmbowman commented:
Sure, just two last questions which recently occurred to me: first, are there ever data URLs (e.g. from matlab) in the content being sanitized by ORA or the Recommender XBlock?  We were mutating the global list of accepted protocols before, whereas now we're just customizing the list in this one place, so the calls to bleach in those two packages will now probably strip any data URLs.  I think the new way is better, just wondering if we need to match it in those other calls for backwards compatibility.

Second, bleach 1.5 significantly pruned the list of protocols allowed by default.  The new code will strip out URLs using mailto, ftp, etc.; are we sure that the loss of those won't be a problem?

ahmedaljazzar commented:
> are there ever data URLs (e.g. from matlab) in the content being sanitized by ORA or the Recommender XBlock? We were mutating the global list of accepted protocols before, whereas now we're just customizing the list in this one place, so the calls to bleach in those two packages will now probably strip any data URLs. I think the new way is better, just wondering if we need to match it in those other calls for backwards compatibility.

Good point, @jmbowman. So actually for ORA and the RecommenderXblock all the calls to `bleach` are backward compatible since upgrading the library version in requirements was enough to keep the same behavior we are expecting.  
However for edX Platform, the other alternative is to store the safe configuration in settings and to make a single point of entry to bleach which is counter-intuitive in my opinion and this might strict some other usages for different Blocks and will increase the possibility of opening protocols not needed for some services to have some functionality running in another one. So my suggestion is to keep using bleach library and whenever we want to modify the default protocols for some feature that will go to code review every time and will be specific to that feature.

> Second, bleach 1.5 significantly pruned the list of protocols allowed by default. The new code will strip out URLs using mailto, ftp, etc.; are we sure that the loss of those won't be a problem?

Yes, that won't be a problem as it's been tested in the new [HTML XBlock](https://github.com/open-craft/xblock-html) and it didn't affect the rendering on LMS nor CMS.

jmbowman commented:
Ok, the ora2 and RecommenderXBlock PRs have been merged.  I think ora2 will need a version bump and a manual upload to PyPI for the new version requirement to take effect (since it depends on an exact version of bleach; ora2 hasn't been updated to comply with OEP-18 yet).  Do you want to prepare the PR for that?

ahmedaljazzar commented:
Sure @jmbowman, I'll do that.

ahmedaljazzar commented:
@jmbowman I bumped the version on ORA2 and the Recommender XBlock, but I think uploading the version to PyPi might require some access, right?

jmbowman commented:
I just noticed that we haven't uploaded either of those packages to PyPI yet...sigh.  For now you should be good just updating the GitHub URLs in `requirements/edx/github.in` and running `make upgrade`.

ahmedaljazzar commented:
@jmbowman I updated the requirements, squashed the commits, rebased the branch on the current master, and currently waiting for the tests results. However, running `make upgrade` in the LMS shell fails with the following output:

```
pip-compile --no-emit-trusted-host --no-index --upgrade -o requirements/edx/testing.txt requirements/edx/testing.in
No handlers could be found for logger "pip.vcs.git"
Could not find a version that matches urllib3<1.24,<1.25,<2.0,==1.24,>=1.21.1,>=1.8
Tried: 0.3, 0.3, 1.0, 1.0, 1.0.1, 1.0.1, 1.0.2, 1.0.2, 1.1, 1.1, 1.2, 1.2, 1.2.1, 1.2.1, 1.2.2, 1.2.2, 1.3, 1.3, 1.4, 1.4, 1.5, 1.5, 1.6, 1.6, 1.7, 1.7, 1.7.1, 1.7.1, 1.8, 1.8, 1.8.2, 1.8.2, 1.8.3, 1.8.3, 1.9, 1.9, 1.9.1, 1.9.1, 1.9.1, 1.9.1, 1.10, 1.10, 1.10, 1.10, 1.10.1, 1.10.1, 1.10.1, 1.10.1, 1.10.2, 1.10.2, 1.10.2, 1.10.2, 1.10.3, 1.10.3, 1.10.3, 1.10.3, 1.10.4, 1.10.4, 1.10.4, 1.10.4, 1.11, 1.11, 1.11, 1.11, 1.12, 1.12, 1.12, 1.12, 1.13, 1.13, 1.13, 1.13, 1.13.1, 1.13.1, 1.13.1, 1.13.1, 1.14, 1.14, 1.14, 1.14, 1.15, 1.15, 1.15, 1.15, 1.15.1, 1.15.1, 1.15.1, 1.15.1, 1.16, 1.16, 1.16, 1.16, 1.17, 1.17, 1.17, 1.17, 1.18, 1.18, 1.18, 1.18, 1.18.1, 1.18.1, 1.18.1, 1.18.1, 1.19, 1.19, 1.19, 1.19, 1.19.1, 1.19.1, 1.19.1, 1.19.1, 1.20, 1.20, 1.20, 1.20, 1.21, 1.21, 1.21, 1.21, 1.21.1, 1.21.1, 1.21.1, 1.21.1, 1.22, 1.22, 1.22, 1.22, 1.23, 1.23, 1.23, 1.23, 1.24, 1.24, 1.24, 1.24
There are incompatible versions in the resolved dependencies.
```

So I went to `requirements/edx/testing.in` but couldn't identify which package is causing this conflict yet. Will be investigating this while the tests are running.

jmbowman commented:
If you add `-v` or `--verbose` to the pip-compile command in the `Makefile`'s upgrade target, you should get some additional information on which packages imposed each of those constraints on the `urllib` version.

ahmedaljazzar commented:
So the problem is that `elasticsearch==1.9.0` requires `urllib3<2.0,>=1.8` while `requests==2.20.0`         requires `urllib3<1.25,>=1.21.1`. So I'm not sure whether to pin one of these packages in this case or to ignore the upgrade in this PR.

jmbowman commented:
It looks like this stems from a recent clash between the latest `urllib3` release and `requests`; see https://github.com/transifex/transifex-client/pull/241 and http://docs.python-requests.org/en/master/community/updates/#id1 .  We'll probably need to add the same restriction to `paver.in` on `urllib3` as exists in `transifex-client` until they update it to account for the latest `requests` release that restored compatibility.

ahmedaljazzar commented:
@jmbowman The packages upgrade went successfully, I also rebased the PR on top of master to check for the quality error that keeps failing on quality errors I didn't include in this PR, still checking to see where the problem is.

jmbowman commented:
Sorry about that failure being hard to track down and there being some inaccuracies in the final job results; I'm filing a couple of tickets to fix that.  The underlying failure is in https://build.testeng.edx.org/job/edx-platform-test-subset/1477550/artifact/reports/xsscommitlint/xsscommitlint.report .  It actually isn't anything you changed, just a preexisting problem with one of the files you modified.

So this triggered a warning because it uses a code pattern that can often enable a cross-site scripting (XSS) vulnerability, but in this case the generated string is only used to find a substring from existing HTML (and not inserted into the rendered output).  You should be ok just whitelisting this particular instance of the warning by adding the following comment immediately before that line:

```python
# xss-lint: disable=python-interpolate-html
```

More docs on the XSS linter can be found at https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/preventing_xss.html

jmbowman commented:
Man, you're hitting all the fun problems.  The new version of `pycodestyle` introduced some new checks, including the "W504" one that's showing up in that quality failure.  It's supposed to be disabled by default, but that default list is ignored due to our configuration in setup.cfg: https://github.com/edx/edx-platform/blob/master/setup.cfg#L33 .  You can add "W504" to that setting to suppress those.  That version of pycodestyle had been out for months, but flake8 didn't allow using it until a new version of that came out a few days ago.

The bok-choy failure looks like a random occurrence, it'll probably pass on the next commit; I'm flagging that for possible removal, it looks like it's failing in about 2% of the test runs.

Let me know if I should do a separate PR just to get us into a good state regarding the dependencies unrelated to your change; normally we don't hit this many issues in a single PR, but it looks like nobody had tried updating dependencies in a couple of weeks.

ahmedaljazzar commented:
@jmbowman I have to admit, this was really challenging to track and I couldn't have been able to fix the tests without your help. Thanks so much.
I based these changes on the new master which includes the fix for `W504` you mentioned above. I also upgraded the requirements as requested.

This is now ready for your action.

edx-webhook commented:
@ahmedaljazzar 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/19063&entry.1484655318=https://github.com/ahmedaljazzar&entry.752974735=2018-10-04+17:42&entry.1917517419=2018-11-01+13:47&entry.2133058324=Yes) so we can improve your experience in the future.

jmbowman commented:
Thanks for persevering with this!  Already fixed one of the problems you hit, with a couple of other improvements coming soon.
